File: ts/src/grpc-tx-indexer.ts
/**
 * Indexes transactions off a Yellowstone gRPC stream, isolating v1.
 *
 * `SubscribeRequestFilterTransactions` has no version field, so a consumer
 * subscribes to everything and discriminates client-side.
 *
 * Run with `just ts-grpc-tx-indexer`.
 */

import { getBase58Decoder, getBlockhashDecoder } from '@solana/kit';

import { computeBudgetOfMessage, formatComputeBudget } from './lib/budget';
import {
    ALL_TRANSACTIONS_REQUEST,
    DEFAULT_GRPC_ENDPOINT,
    exitOnStreamClose,
    messageVersion,
    readEnvLimit,
    subscribe,
    type SubscribeUpdate,
} from './lib/grpc';

const base58 = getBase58Decoder();
const blockhash = getBlockhashDecoder();

const exitAfterV1 = readEnvLimit('TXV1_EXIT_AFTER_V1');
const { close, stream } = await subscribe(DEFAULT_GRPC_ENDPOINT, ALL_TRANSACTIONS_REQUEST);
const tally = { legacy: 0, v0: 0, v1: 0 };

console.log(`subscribed to ${DEFAULT_GRPC_ENDPOINT}`);

stream.on('data', (update: SubscribeUpdate) => {
    const message = update.transaction?.transaction?.transaction?.message;
    if (!message) {
        return;
    }

    const version = messageVersion(message);
    tally[version] += 1;

    // This is the line an existing indexer has to change: scanning
    // ComputeBudget instructions alone returns nothing for v1 without erroring.
    const budget = computeBudgetOfMessage(message);
    const signature = base58.decode(update.transaction!.transaction!.signature);
    console.log(
        `slot ${update.transaction!.slot} ${version} ` +
            `cuLimit=${budget.computeUnitLimit ?? '-'} ` +
            `priorityFee=${budget.priorityFeeLamports ?? '-'} lamports sig ${signature}`,
    );

    if (version !== 'v1') {
        return;
    }

    console.log(formatComputeBudget(budget));
    // On v1 the wire's `recentBlockhash` slot carries the lifetime specifier.
    console.log(`  lifetimeSpecifier:            ${blockhash.decode(message.recentBlockhash)}`);
    console.log(`  addressTableLookups:          ${message.addressTableLookups.length} (v1 never has any)`);
    console.log(`  running tally:                legacy=${tally.legacy} v0=${tally.v0} v1=${tally.v1}`);

    if (exitAfterV1 > 0 && tally.v1 >= exitAfterV1) {
        console.log('\nreached the v1 limit, exiting');
        close();
        process.exit(0);
    }
});

exitOnStreamClose(stream, close);

File: rust/src/bin/grpc-tx-indexer.rs
//! Indexes transactions off a Yellowstone gRPC stream, isolating v1.
//!
//! `SubscribeRequestFilterTransactions` has no version field, so a consumer
//! subscribes to everything and discriminates on the presence of `config`.

use anyhow::Result;
use clap::Parser;
use futures::StreamExt;
use std::collections::HashMap;
use txv1::grpc::{describe_config, message_version};
use txv1::version::VersionTally;
use txv1::{ComputeBudget, MessageVersion, DEFAULT_GRPC_URL};
use yellowstone_grpc_client::GeyserGrpcClient;
use yellowstone_grpc_proto::geyser::{
    subscribe_update::UpdateOneof, CommitmentLevel, SubscribeRequest,
    SubscribeRequestFilterTransactions,
};

#[derive(Parser)]
#[command(about = "Stream transactions over gRPC and decode v1 configs")]
struct Args {
    #[arg(long, env = "TXV1_GRPC_URL", default_value = DEFAULT_GRPC_URL)]
    grpc_url: String,

    /// Stop once this many v1 transactions have been seen.
    #[arg(long)]
    exit_after_v1: Option<u64>,
}

#[tokio::main]
async fn main() -> Result<()> {
    let args = Args::parse();
    let endpoint = args.grpc_url.clone();
    let mut client = GeyserGrpcClient::build_from_shared(args.grpc_url)?
        .connect()
        .await?;

    let request = SubscribeRequest {
        commitment: Some(CommitmentLevel::Confirmed as i32),
        transactions: HashMap::from([(
            "all".to_string(),
            SubscribeRequestFilterTransactions {
                failed: Some(false),
                vote: Some(false),
                ..Default::default()
            },
        )]),
        ..Default::default()
    };

    let mut stream = client.subscribe_once(request).await?;
    let mut tally = VersionTally::default();
    println!("subscribed to {endpoint}");

    while let Some(update) = stream.next().await {
        let Some(UpdateOneof::Transaction(update)) = update?.update_oneof else {
            continue;
        };
        let Some(info) = update.transaction else {
            continue;
        };
        let signature = bs58::encode(&info.signature).into_string();
        let Some(message) = info.transaction.and_then(|tx| tx.message) else {
            continue;
        };

        let version = message_version(&message);
        tally.record(version);

        // This is the line an existing indexer has to change: scanning
        // ComputeBudget instructions alone returns nothing for v1 without
        // erroring.
        let budget = ComputeBudget::of_message(&message);
        println!(
            "slot {} {version} cu_limit={} priority_fee={} lamports sig {signature}",
            update.slot,
            budget
                .compute_unit_limit
                .map_or_else(|| "-".to_string(), |limit| limit.to_string()),
            budget
                .priority_fee_lamports
                .map_or_else(|| "-".to_string(), |fee| fee.to_string()),
        );

        if version != MessageVersion::V1 {
            continue;
        }

        if let Some(config) = &message.config {
            println!("{}", describe_config(config, "  "));
        }
        // On v1 the wire's `recent_blockhash` slot carries the lifetime
        // specifier.
        println!(
            "  lifetime_specifier:              {}",
            bs58::encode(&message.recent_blockhash).into_string()
        );
        println!(
            "  address_table_lookups:           {} (v1 never has any)",
            message.address_table_lookups.len()
        );
        println!("  running tally:                   {tally}");

        if args.exit_after_v1.is_some_and(|limit| tally.v1 >= limit) {
            println!("\nreached the v1 limit, exiting");
            break;
        }
    }

    Ok(())
}
